File: Services/Medico.Api/ClientApp/src/app/_models/patientChartNode.ts
import { PatientChartNodeType } from './patientChartNodeType';
import { Template } from './template';
import { PatientChartNodeAttributes } from './patientChartNodeAttributes';
import { PatientChartNodeTemplateProviderService } from '../_services/patient-chart-node-template-provider.service';

export class PatientChartNode {
  id!: string;
  name!: string;
  title!: string;
  type!: PatientChartNodeType;
  value?: any | any[];
  attributes!: PatientChartNodeAttributes;
  children?: PatientChartNode[];
  parentId?: string;
  template?: string;

  static createPatientChartNode(
    id: string,
    name: string,
    title: string,
    type: PatientChartNodeType,
    value: any | any[],
    attributes: PatientChartNodeAttributes,
    parentId: string | undefined,
    template: string
  ): PatientChartNode {
    // console.log("in createPatientChartNode")

    const patientChartNode = new PatientChartNode();
    patientChartNode.id = id;
    patientChartNode.name = name;
    patientChartNode.title = title;
    patientChartNode.type = type;
    patientChartNode.value = value;
    patientChartNode.attributes = attributes;
    patientChartNode.children = [];
    patientChartNode.parentId = parentId;
    patientChartNode.template = template;

    return patientChartNode;
  }

  static createPatientChartTemplateNode(
    id: string,
    parentId: string | undefined,
    template: Template,
    templateTypeName: string
  ): PatientChartNode {
    // console.log("in createPatientChartTemplateNode")

    const patientChartNode = new PatientChartNode();

    patientChartNode.id = id;
    patientChartNode.name = template.reportTitle;
    patientChartNode.title = template.reportTitle;
    patientChartNode.type = PatientChartNodeType.TemplateNode;
    patientChartNode.value = {
      defaultTemplateHtml: template.defaultTemplateHtml,
      detailedTemplateHtml: template.initialDetailedTemplateHtml,
      isDetailedTemplateUsed: !template.defaultTemplateHtml,
    };

    patientChartNode.parentId = parentId;
    patientChartNode.template =
      PatientChartNodeTemplateProviderService.getTemplateValueForPatientChartTemplateNode(
        template.id,
        templateTypeName
      );

    const nodeSpecificAttributes = {
      templateId: template.id,
    };

    const attributes = PatientChartNodeAttributes.createPatientChartNodeAttributes(
      template.templateOrder as number,
      true,
      false,
      false,
      false,
      nodeSpecificAttributes
    );

    patientChartNode.attributes = attributes;

    return patientChartNode;
  }

  static createPatientChartTemplateListNode(
    id: string,
    name: string,
    title: string,
    attributes: PatientChartNodeAttributes,
    parentId?: string
  ): PatientChartNode {
    // console.log("in createPatientChartTemplateListNode")

    const patientChartNode = new PatientChartNode();
    patientChartNode.id = id;
    patientChartNode.name = name;
    patientChartNode.title = title;
    patientChartNode.type = PatientChartNodeType.TemplateListNode;
    patientChartNode.value = [];
    patientChartNode.attributes = attributes;
    patientChartNode.children = [];
    patientChartNode.parentId = parentId;
    patientChartNode.template =
      PatientChartNodeTemplateProviderService.getTemplateValueForPatientChartTemplateListNode(
        name
      );

    return patientChartNode;
  }

  static createPatientChartSectionNode(
    id: string,
    name: string,
    title: string,
    attributes: PatientChartNodeAttributes,
    parentId: string | undefined,
    chartId: string
  ): PatientChartNode {
    // console.log("in createPatientChartSectionNode");
    // console.log("\t" + title);

    const patientChartNode = new PatientChartNode();
    patientChartNode.id = id;
    patientChartNode.name = name;
    patientChartNode.title = title;
    patientChartNode.type = PatientChartNodeType.SectionNode;
    patientChartNode.value = {
      chartId,
    };
    patientChartNode.attributes = attributes;
    patientChartNode.children = [];
    patientChartNode.parentId = parentId;
    patientChartNode.template = '';

    return patientChartNode;
  }
}

File: Application/Medico.Application/Services/PatientChart/PatientChartNode.cs
using System;
using System.Collections.Generic;
using Medico.Application.ViewModels.Template;

namespace Medico.Application.Services.PatientChart
{
    public class PatientChartNode
    {
        public Guid Id { get; set; }

        public string Name { get; set; }

        public string Title { get; set; }

        public PatientChartNodeType Type { get; set; }

        public dynamic Value { get; set; }

        public PatientChartNodeAttributes Attributes { get; set; }

        public List<PatientChartNode> Children { get; set; }

        public Guid ParentId { get; set; }

        public string Template { get; set; }

        public static PatientChartNode CreatePatientChartTemplateNode(Guid id,
            PatientChartNode parentNode, TemplateVm template, string templateTypeName)
        {
            var templateOrder = template.TemplateOrder;
            var templateId = template.Id;

            return new PatientChartNode
            {
                Id = id,
                Name = template.ReportTitle,
                Title = template.ReportTitle,
                Type = PatientChartNodeType.TemplateNode,
                Value = new
                {
                    DefaultTemplateHtml = template.DefaultTemplateHtml,
                    DetailedTemplateHtml = template.InitialDetailedTemplateHtml,
                    IsDetailedTemplateUsed = string.IsNullOrEmpty(template.DefaultTemplateHtml)
                },
                ParentId = parentNode.Id,
                Template = PatientChartNodeTemplates
                    .GetTemplateValueForPatientChartTemplateNode(templateId, templateTypeName),
                Attributes =
                    PatientChartNodeAttributes.CreatePatientChartNodeAttributes(templateOrder,
                        true, false, false, false, parentNode.Attributes.ResponsibleEmployeeTypes
                        , parentNode.Attributes.ChartColors, parentNode.Attributes.AuditRequired
                        , new NodeSpecificAttributes { TemplateId = templateId })
            };
        }
    }
}